service/cluster/cluster.go

```diff
@@ -0,0 +1,356 @@
+	}
+
+	httpClient := &http.Client{
+		Transport: &http2.Transport{
+			AllowHTTP: true,
+			DialTLSContext: func(ctx context.Context, network, addr string, _ *tls.Config) (net.Conn, error) {
+				// Set a maximum timeout for establishing the connection. If our
+				// context has a deadline earlier than our timeout, we shrink the
+				// timeout to it.
+				//
+				// TODO(rfratto): consider making the max timeout configurable.
+				timeout := 30 * time.Second
+				if dur, ok := deadlineDuration(ctx); ok && dur < timeout {
+					timeout = dur
+				}
+
+				return net.DialTimeout(network, addr, timeout)
+			},
+		},
+	}
+
+	node, err := ckit.NewNode(httpClient, ckitConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cluster node: %w", err)
+	}
+	if opts.Metrics != nil {
+		opts.Metrics.MustRegister(node.Metrics())
+	}
+
+	return &Service{
+		log:    l,
+		tracer: t,
+		opts:   opts,
+
+		sharder: ckitConfig.Sharder,
+		node:    node,
+	}, nil
+}
+
+func deadlineDuration(ctx context.Context) (d time.Duration, ok bool) {
+	if t, ok := ctx.Deadline(); ok {
+		return time.Until(t), true
+	}
+	return 0, false
+}
+
+// Definition returns the definition of the cluster service.
+func (s *Service) Definition() service.Definition {
+	return service.Definition{
+		Name:       ServiceName,
+		ConfigType: nil, // cluster does not accept configuration.
+		DependsOn: []string{
+			// Cluster depends on the HTTP service to work properly.
+			http_service.ServiceName,
+		},
+	}
+}
+
+// ServiceHandler returns the service handler for the clustering service. The
+// resulting handler always returns 404 when clustering is disabled.
+func (s *Service) ServiceHandler(host service.Host) (base string, handler http.Handler) {
+	base, handler = s.node.Handler()
+
+	if !s.opts.EnableClustering {
+		handler = http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+			http.Error(w, "clustering is disabled", http.StatusNotFound)
+		})
+	}
+
+	return base, handler
+}
+
+// ChangeState changes the state of the service. If clustering is enabled,
+// ChangeState will block until the state change has been propagated to another
+// node; cancel the current context to stop waiting. ChangeState fails if the
+// current state cannot move to the provided targetState.
+//
+// Note that the state must be StateParticipant to receive writes.
+func (s *Service) ChangeState(ctx context.Context, targetState peer.State) error {
+	return s.node.ChangeState(ctx, targetState)
+}
+
+// Run starts the cluster service. It will run until the provided context is
+// canceled or there is a fatal error.
+func (s *Service) Run(ctx context.Context, host service.Host) error {
+	// Stop the node on shutdown.
+	defer s.stop()
+
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	s.node.Observe(ckit.FuncObserver(func(peers []peer.Peer) (reregister bool) {
+		if ctx.Err() != nil {
+			return false
+		}
```

Comment:
Is this and the other call to `ctx.Err()` below to avoid calling a NotifyClusterChange once we're setting our state to terminating, or just to clean up the Observer once the parent context is exiting?

Reply:
Two reasons:

* We allow ourselves to clean up the Observer once the cluster service is no longer running.
* We allow ourselves to stop early instead of forcing iterating over all components so we can terminate sooner.

Do you want me to add this as comments?

Reply:
Yeah, it wouldn't hurt!